addons/website_sale/models/sale_order.py

```diff
@@ -190,7 +190,9 @@ def _cart_update(self, product_id, line_id=None, add_qty=0, set_qty=0, **kwargs)
 
         order_line = self._cart_update_order_line(product_id, quantity, order_line, **kwargs)
 
-        if order_line and order_line.price_unit == 0 and self.website_id.prevent_zero_price_sale:
+        if order_line and order_line.price_unit == 0 and \
+                self.website_id.prevent_zero_price_sale and \
+                product.detailed_type not in ['event', 'event_booth', 'course']:
```

Comment:
I think that we could have a model method to get the list of white listed product types and update it in each module when we need to. I don't think that updating website_sale each time when we want to update that list is good. It will be a bit more verbose but at least we avoid to add stuff not related to the website_sale module. And it could perhaps be useful for custom modules.